spatialos-sdk/src/worker/worker_future.rs

```diff
@@ -0,0 +1,124 @@
+    pub(crate) shared_result: Arc<Mutex<Option<T::Output>>>,
+    pub(crate) thread: Option<JoinHandle<()>>,
+}
+
+impl<T: WorkerSdkFuture> Clone for WorkerFutureHandle<T> {
+    fn clone(&self) -> Self {
+        WorkerFutureHandle {
+            ptr: self.ptr,
+            shared_result: self.shared_result.clone(),
+            thread: None,
+        }
+    }
+}
+
+// SAFE: It should be safe to send WorkerFutureHandle<T> between threads given the following
+// conditions around *mut T::RawPointer.
+//
+// First, some context. For the Worker SDK futures, there are only two operations you can
+// perform with this pointer:
+//
+//  * Poll (blocking or with timeout)
+//  * Destroy
+//
+// Given the following rules, we can guarantee that its thread-safe:
+//
+//  1. Poll cannot be called after Destroy.
+//      We ensure this cannot be true as we store a reference to the handle of the thread where Poll
+//      could be running. If this reference is present in the handle during drop(), we join that thread.
+//      This guarantees us that poll has already begun (and then finished) by the time Destroy is called.
+//      Since we only ever call Poll once, we cannot possibly call Poll after Destroy.
+//  2. Destroy is called at most once.
+//      We call Destroy in the drop() of the WorkerFuture<T> only. Something cannot be dropped more than once.
+//
+// This has the side effect the the Drop impl can block, but this was the already the case anyway with the
+// Worker SDK future Destroy methods (which I believe wait for the future to complete).
+unsafe impl<T: WorkerSdkFuture> Send for WorkerFutureHandle<T> {}
+
+impl<T: WorkerSdkFuture> WorkerFutureHandle<T> {
+    pub(crate) fn new(ptr: *mut T::RawPointer) -> Self {
+        WorkerFutureHandle {
+            ptr,
+            shared_result: Arc::new(Mutex::new(None)),
+            thread: None,
+        }
+    }
+}
+
+pub trait WorkerSdkFuture: Unpin + Send {
+    type RawPointer;
+    type Output;
+
+    fn start(&self) -> *mut Self::RawPointer;
+    unsafe fn get(ptr: *mut Self::RawPointer) -> Self::Output;
+    unsafe fn destroy(ptr: *mut Self::RawPointer);
+}
+
+impl<T: WorkerSdkFuture + 'static> Future for WorkerFuture<T> {
+    type Output = T::Output;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut next_state = None;
+        let mut result = Poll::Pending;
+
+        match self.as_mut().get_mut() {
+            WorkerFuture::NotStarted(future) => {
+                let mut handle = WorkerFutureHandle::new(future.start());
+
+                let thread_handle = handle.clone();
+                let thread_waker = cx.waker().clone();
+
+                handle.thread = Some(thread::spawn(move || unsafe {
+                    let value = T::get(thread_handle.ptr);
+                    thread_handle.shared_result.lock().unwrap().replace(value);
+                    thread_waker.wake();
+                }));
```

Comment:
What sort of options would you want to offer to the user?

Reply:
They might want to have the blocking logic run on their own thread pool instead of a newly-spawned thread. It might also be good to offer the option to offer polling with a timeout instead of only blocking until the future resolves, e.g. polling at 1ms intervals so that the future can be run on any executor without blocking.

Reply:
Yeah I'll open an issue to track this :)

Reply:
https://github.com/jamiebrynes7/spatialos-sdk-rs/issues/137